<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Module 0: Introduction</title>
    <meta charset="utf-8" />
    <meta name="author" content="Ian McCarthy | Emory University" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis-fonts.css" rel="stylesheet" />
    <script src="libs/fabric/fabric.min.js"></script>
    <link href="libs/xaringanExtra-scribble/scribble.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-scribble/scribble.js"></script>
    <script>document.addEventListener('DOMContentLoaded', function() { window.xeScribble = new Scribble({"pen_color":["#E68080"],"pen_size":3,"eraser_size":30,"palette":[]}) })</script>
    <link rel="stylesheet" href="custom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Module 0: Introduction
## <html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=0px>
</html>
### Ian McCarthy | Emory University
### Econ 470 &amp; HLTH 470

---


&lt;!-- Adjust some CSS code for font size, maintain R code font size --&gt;
&lt;style type="text/css"&gt;
.remark-slide-content {
    font-size: 30px;
    padding: 1em 2em 1em 2em;    
}
.remark-code, .remark-inline-code { 
    font-size: 20px;
}
&lt;/style&gt;


&lt;!-- Set R options for how code chunks are displayed and load packages --&gt;





# Table of contents

1. [Motivation](#welcome)

2. [Syllabus highlights](#syllabus)

3. [Software installation](#software)

4. [Starting with Git and RStudio](#git_rstudio)

5. [Tidy Data](#tidy)

6. [Real World](#real_data)

7. [Medicare Advantage](#ma_data)


&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: welcome

# Motivation

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;

---
# Motivating question

**Does health care spending improve health?**

???
Let's motivate the course with this simple question...does health care spending improve our health?

---
# Spending and Health




```r
ggplot(data = (dartmouth.data %&gt;% filter(Year==2015)), 
       mapping = aes(x = Expenditures, y = Total_Mortality)) + 
  geom_point(size = 1) + theme_bw() + scale_x_continuous(label = comma) +
  geom_smooth(method="lm", se=FALSE, color="blue", size=1/2) +
  labs(x = "Spending Per Capita ($US)",
       y = "Mortality Rate",
       title = "Mortality and Health Care Spending")
```
.plot-callout[
&lt;img src="00-introduction_files/figure-html/dartmouth-callout-1.png" style="display: block; margin: auto;" /&gt;
]


---
# Spending and Health

&lt;img src="00-introduction_files/figure-html/dartmouth-output-1.png" style="display: block; margin: auto;" /&gt;


???
This shows a positive relationship between health care spending and mortality. So health care spending is actually negatively correlated with health. What does this mean?

---
# Spending and Health

- Does medical spending make us sicker?
- What else might explain this relationship?

.plot-callout[
&lt;img src="00-introduction_files/figure-html/dartmouth-callout-1.png" style="display: block; margin: auto;" /&gt;
]

???
Sicker people are more likely to need health care. So in this simple graph, we're actually picking up the fact that sicker people are more likely to die. In econometrics, this problem goes by a few names, such as endogeneity or selection.

---
# Goals of this course.

1. Understand and implement selected methods for causal inference
--
&lt;br&gt;
2. Along the way...data management and version control with real data

--
&lt;br&gt;
3. Summarize, visualize, and explain research results


&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: syllabus

# Syllabus highlights

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;

(Read the full document [here](https://imccart.github.io/econ470-content/syllabus-spring2022/syllabus.pdf).)

---
# Why this course?

1. Major problems that need solutions

2. Need good, convincing empirical work for policy

3. Working with data is hard, particularly health care data

4. Your work should be transparent and reproducible

---
# Structure

- Very applied in nature

- *Methods* for causal inference

  - Selection on observables (regression, re-weighting, matching, propensity scores)
  
  - Instrumental variables
  
  - Regression discontinuity
  
  - Difference-in-differences


---
# Structure

- *Substantive* areas

  - Hospital pricing, policy, and competition
  
  - Cigarette taxes and demand
  
  - Medicare Advantage and quality disclosure
  
  - Medicaid expansion and health insurance


---
# Structure

- *Datasets* from the real world

  - Hospital Cost Report Information System (HCRIS)
  
  - Centers for Disease Control (CDC)
  
  - Medicare Advantage data
  
  - Behavioral Risk Factor Surveillance System (BRFSS), Medicaid, Health Insurance Exchanges
  
  
---
# Assignments

- Homework (x5)

- Research project

- Participation

---
# Grading

| Component                           | Weight |
|:-|-:|
| 5 × homework assignments (11% each) | 55% |
| Research project                		| 40% |
| Participation  						          | 5%  |



&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: software

# Software Installation
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;


---
# Software Installation

1. Download [R](https://www.r-project.org/)

2. Download [RStudio](https://www.rstudio.com/products/rstudio/download/preview/)

3. Download [Git](https://git-scm.com/downloads)

4. Create an account on [GitHub](https://github.com/)

--

For help and troubleshooting with Git and GitHub, take a look at Jenny Bryan's http://happygitwithr.com.


---
# Checklist

&lt;html&gt;&amp;#9745;&lt;/html&gt; Do you have the most recent version of R?
  
  ```r
  version$version.string
  ```
  
  ```
  ## [1] "R version 4.1.2 (2021-11-01)"
  ```


&lt;html&gt;&amp;#9745;&lt;/html&gt; Do you have the most recent version of RStudio? (The [preview version](https://www.rstudio.com/products/rstudio/download/preview/) is fine.)
  
  ```r
  RStudio.Version()$version
  ## Requires an interactive session but should return something like "[1] ‘1.4.1717’"
  ```

&lt;html&gt;&amp;#9745;&lt;/html&gt; Have you updated all of your R packages? 
  
  ```r
  update.packages(ask = FALSE, checkBuilt = TRUE)
  ```
  
---
# Checklist

- Open up the [shell](http://happygitwithr.com/shell.html#shell)

- Windows users, make sure that you installed a Bash-compatible version of the shell. If you installed [Git for Windows](https://gitforwindows.org), then you should be good to go.


---
# Checklist

&lt;html&gt;&amp;#9745;&lt;/html&gt; Which version of Git have you installed?
  
  ```r
  git --version
  ```

&lt;html&gt;&amp;#9745;&lt;/html&gt; Did you introduce yourself to Git? (Substitute in your details.)
  
  ```r
  git config --global user.name 'Ian McCarthy'
  git config --global user.email 'ian.mccarthy@emory.edu'
  git config --global --list
  ```

&lt;html&gt;&amp;#9745;&lt;/html&gt; Did you register an account in GitHub?

---
# Altnernative setup...

Just use the cloud!

- We have our own virtual computer via AWS
- This computer has all the space you need and the data are already available
- Downside: the computer will be "on" during designated times only
- Details of this are on Canvas


&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: git_rstudio

# Practice with Git and RStudio
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;


---
# Before next class (see http://happygitwithr.com)

1. Download [R](https://www.r-project.org/)

2. Download [RStudio](https://www.rstudio.com/products/rstudio/download/preview/)

3. Download [Git](https://git-scm.com/downloads)

4. Create an account on [GitHub](https://github.com/)

5. Connect RStudio to Git and GitHub

6. Start/clone/fork a repository for this class 


---
# Setting things up

Now we're going to clone a GitHub repository (repo) using RStudio. The video below is from Grant McDermott's class.

&lt;center&gt;
&lt;iframe width="600" height="350" src="https://raw.githack.com/uo-ec607/lectures/master/02-git/02-Git.html#15" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/center&gt;

---
# Some common mistakes for windows users

- Windows folders are *not* files...there is no content without a file. You can't commit or push changes without content.

- Let RStudio/GitHub create the directory (main folder) for you.

- If you're working across devices on your own repo, be sure to pull before starting and push afterward.

- Avoid spaces in file names. Avoid them at all costs. *DO NOT PUT SPACES IN YOUR FILE NAMES.*

&gt;*"A space in a file name is a space in your soul."*


---
# Ideal workflow

Until you are a Git(Hub) expert...

1. Start project on GitHub (fork from another repo if needed)

2. Clone to desktop with RStudio

3. See http://happygitwithr.com for instructions on linking your local repo with a new upstream remote



---
# Working with AWS

Let's do this live!


&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: tidy

# Tidy Data

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;


---
# The tidyverse

- Suite of packages collectively known as the [tidyverse](https://www.tidyverse.org/)
- Different from `base` R in many ways
- The tidyverse with pipes&lt;sup&gt;1&lt;/sup&gt; is more intuitive to me

.footnote[
&lt;sup&gt;1&lt;/sup&gt; We'll talk about pipes very soon!
] 


---
# What is Tidy data?

Resources:
- Paper: [*Tidy Data*](https://vita.had.co.nz/papers/tidy-data.pdf) (Hadley Wickham, 2014 JSS)
- Vignette: [Tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html) (from the `tidyr` package)

--
&lt;br&gt;
&lt;br&gt;

Essentially:
1. Variables are columns
2. Observations are rows
3. Variables and observations make a table


---
# Intro to Tidy data

Let's load the tidyverse package and check the output:

```r
library(tidyverse)
```

--

Comes with lots of other packages like `ggplot2`, `tibble`, `dplyr`, etc.


---
# Pipes: %&gt;%

- The pipe operator is denoted `%&gt;%` and is automatically loaded with the tidyverse.
  
- Pipes are awesome!


```r
## These next two lines of code do exactly the same thing.
mpg %&gt;% filter(manufacturer=="audi") %&gt;% group_by(model) %&gt;% summarise(hwy_mean = mean(hwy))
summarise(group_by(filter(mpg, manufacturer=="audi"), model), hwy_mean = mean(hwy))
```

--

The first line reads from left to right and from data to operation. The `Base R` version (line 2) works in the opposite order.

---
# Pipes: %&gt;%

Helps to break the pipes over several lines

```r
mpg %&gt;% 
  filter(manufacturer=="audi") %&gt;% 
  group_by(model) %&gt;% 
  summarise(hwy_mean = mean(hwy))
```

```
## # A tibble: 3 × 2
##   model      hwy_mean
##   &lt;chr&gt;         &lt;dbl&gt;
## 1 a4             28.3
## 2 a4 quattro     25.8
## 3 a6 quattro     24
```

---
# The dplyr package:

1. `filter()`: Find or exclude certian rows

2. `arrange()`: Sort your observations

3. `select()`: Select specific variables

4. `mutate()`: Create new variables

5. `summarise()`: Collapse multiple rows into a single summary value


---
# 1) dplyr::filter()

Multiple filters separated by commas:

```r
starwars %&gt;% 
  filter( 
    species == "Human", 
    height &gt;= 190
    ) %&gt;% head(5)
```

```
## # A tibble: 4 × 14
##   name      height  mass hair_color skin_color eye_color birth_year sex   gender
##   &lt;chr&gt;      &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; 
## 1 Darth Va…    202   136 none       white      yellow          41.9 male  mascu…
## 2 Qui-Gon …    193    89 brown      fair       blue            92   male  mascu…
## 3 Dooku        193    80 white      fair       brown          102   male  mascu…
## 4 Bail Pre…    191    NA black      tan        brown           67   male  mascu…
## # … with 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;, films &lt;list&gt;,
## #   vehicles &lt;list&gt;, starships &lt;list&gt;
```

---
# 1) dplyr::filter()

Common to use `filter()` to check or remove missing values

```r
starwars %&gt;% 
  filter(is.na(height)) %&gt;% head(5)
```

```
## # A tibble: 5 × 14
##   name     height  mass hair_color skin_color eye_color birth_year sex   gender 
##   &lt;chr&gt;     &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  
## 1 Arvel C…     NA    NA brown      fair       brown             NA male  mascul…
## 2 Finn         NA    NA black      dark       dark              NA male  mascul…
## 3 Rey          NA    NA brown      light      hazel             NA fema… femini…
## 4 Poe Dam…     NA    NA brown      light      brown             NA male  mascul…
## 5 BB8          NA    NA none       none       black             NA none  mascul…
## # … with 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;, films &lt;list&gt;,
## #   vehicles &lt;list&gt;, starships &lt;list&gt;
```

---
count: false

# 1) dplyr::filter()

To remove missing observations, use `filter(!is.na(height))`. Also try the `drop_na()` function from `tidyr`.

---
# 2) dplyr::arrange()

Arrange in ascending order:

```r
starwars %&gt;% 
  arrange(birth_year) %&gt;% head(5)
```

```
## # A tibble: 5 × 14
##   name      height  mass hair_color skin_color eye_color birth_year sex   gender
##   &lt;chr&gt;      &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; 
## 1 Wicket S…     88    20 brown      brown      brown              8 male  mascu…
## 2 IG-88        200   140 none       metal      red               15 none  mascu…
## 3 Luke Sky…    172    77 blond      fair       blue              19 male  mascu…
## 4 Leia Org…    150    49 brown      light      brown             19 fema… femin…
## 5 Wedge An…    170    77 brown      fair       hazel             21 male  mascu…
## # … with 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;, films &lt;list&gt;,
## #   vehicles &lt;list&gt;, starships &lt;list&gt;
```

---
# 2) dplyr::arrange()

Arrange descending order using `arrange(desc())`:

```r
starwars %&gt;% 
  arrange(desc(birth_year)) %&gt;% head(5)
```

```
## # A tibble: 5 × 14
##   name    height  mass hair_color skin_color  eye_color birth_year sex    gender
##   &lt;chr&gt;    &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; 
## 1 Yoda        66    17 white      green       brown            896 male   mascu…
## 2 Jabba …    175  1358 &lt;NA&gt;       green-tan,… orange           600 herma… mascu…
## 3 Chewba…    228   112 brown      unknown     blue             200 male   mascu…
## 4 C-3PO      167    75 &lt;NA&gt;       gold        yellow           112 none   mascu…
## 5 Dooku      193    80 white      fair        brown            102 male   mascu…
## # … with 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;, films &lt;list&gt;,
## #   vehicles &lt;list&gt;, starships &lt;list&gt;
```

---
# 3) dplyr::select()

- Use commas to select multiple columns 
- Use "first:last" for consecutive columns
- Deselect a column with "-"

---
# 3) dplyr::select()


```r
starwars %&gt;% 
  select(name:skin_color, species, -height) %&gt;% head(5)
```

```
## # A tibble: 5 × 5
##   name            mass hair_color skin_color  species
##   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;  
## 1 Luke Skywalker    77 blond      fair        Human  
## 2 C-3PO             75 &lt;NA&gt;       gold        Droid  
## 3 R2-D2             32 &lt;NA&gt;       white, blue Droid  
## 4 Darth Vader      136 none       white       Human  
## 5 Leia Organa       49 brown      light       Human
```

---
# 3) dplyr::select()

Rename within `select()`:

```r
starwars %&gt;%
  select(alias=name, crib=homeworld, sex=gender)  %&gt;% head(5)
```

```
## # A tibble: 5 × 3
##   alias          crib     sex      
##   &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;    
## 1 Luke Skywalker Tatooine masculine
## 2 C-3PO          Tatooine masculine
## 3 R2-D2          Naboo    masculine
## 4 Darth Vader    Tatooine masculine
## 5 Leia Organa    Alderaan feminine
```

---
# 3) dplyr::select()

Use `select(contains(PATTERN))` to find rows that contain some strings of interest

```r
starwars %&gt;% 
  select(name, contains("color")) %&gt;% head(5)
```

```
## # A tibble: 5 × 4
##   name           hair_color skin_color  eye_color
##   &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;    
## 1 Luke Skywalker blond      fair        blue     
## 2 C-3PO          &lt;NA&gt;       gold        yellow   
## 3 R2-D2          &lt;NA&gt;       white, blue red      
## 4 Darth Vader    none       white       yellow   
## 5 Leia Organa    brown      light       brown
```

--

Also look into the `stringr` package.

---
# 4) dplyr::mutate()

Create new variables with `mutate()`

```r
starwars %&gt;% 
  select(name, birth_year) %&gt;%
  mutate(dog_years = birth_year * 7) %&gt;%
  mutate(comment = paste0(name, " is ", dog_years, " in dog years."))  %&gt;% head(5)
```

```
## # A tibble: 5 × 4
##   name           birth_year dog_years comment                            
##   &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;                              
## 1 Luke Skywalker       19        133  Luke Skywalker is 133 in dog years.
## 2 C-3PO               112        784  C-3PO is 784 in dog years.         
## 3 R2-D2                33        231  R2-D2 is 231 in dog years.         
## 4 Darth Vader          41.9      293. Darth Vader is 293.3 in dog years. 
## 5 Leia Organa          19        133  Leia Organa is 133 in dog years.
```

---
# 4) dplyr::mutate()

`mutate()` processes in order, so you can put dependent mutates one after another:

```r
starwars %&gt;% 
  select(name, birth_year) %&gt;%
  mutate(dog_years = birth_year * 7, ## Separate with a comma
         comment = paste0(name, " is ", dog_years, " in dog years.")) %&gt;% head(5)
```

```
## # A tibble: 5 × 4
##   name           birth_year dog_years comment                            
##   &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;                              
## 1 Luke Skywalker       19        133  Luke Skywalker is 133 in dog years.
## 2 C-3PO               112        784  C-3PO is 784 in dog years.         
## 3 R2-D2                33        231  R2-D2 is 231 in dog years.         
## 4 Darth Vader          41.9      293. Darth Vader is 293.3 in dog years. 
## 5 Leia Organa          19        133  Leia Organa is 133 in dog years.
```

---

# 4) dplyr::mutate()

Other handy ways to use `mutate()`:

```r
starwars %&gt;% 
  select(name, height) %&gt;%
  filter(name %in% c("Luke Skywalker", "Anakin Skywalker")) %&gt;% 
  mutate(tall1 = height &gt; 180) %&gt;%
  mutate(tall2 = ifelse(height &gt; 180, "Tall", "Short")) ## Same effect, but can choose labels
```

```
## # A tibble: 2 × 4
##   name             height tall1 tall2
##   &lt;chr&gt;             &lt;int&gt; &lt;lgl&gt; &lt;chr&gt;
## 1 Luke Skywalker      172 FALSE Short
## 2 Anakin Skywalker    188 TRUE  Tall
```

---
# 4) dplyr::mutate()

Note the "scoped" variants of `mutate()` that work on a subset of variables:
- `mutate_all()` affects every variable
- `mutate_at()` affects named or selected variables
- `mutate_if()` affects variables that meet some criteria (e.g. are numeric)

--

These have since been replaced with `across()`. So you would use `mutate(across())` instead.

---
# 5) dplyr::summarise() with group_by()


```r
starwars %&gt;% 
  group_by(species, gender) %&gt;% 
  summarise(mean_height = mean(height, na.rm = T)) %&gt;% head(5)
```

```
## # A tibble: 5 × 3
## # Groups:   species [5]
##   species  gender    mean_height
##   &lt;chr&gt;    &lt;chr&gt;           &lt;dbl&gt;
## 1 Aleena   masculine          79
## 2 Besalisk masculine         198
## 3 Cerean   masculine         198
## 4 Chagrian masculine         196
## 5 Clawdite feminine          168
```

--
*Note*: `na.rm = T` is usually a good idea, otherwise your summary will be `NA` too.


---
# 5) dplyr::summarise()

"scoped" variants also work with `summarise()`
- `summarise_all()` affects every variable
- `summarise_at()` affects named or selected variables
- `summarise_if()` affects variables that meet some criteria (e.g. are numeric)

--

Again, supersceded by `across()`.

---
count: false

# 5) dplyr::summarise()


```r
starwars %&gt;% group_by(species) %&gt;% summarise(across(where(is.numeric), mean, na.rm=TRUE)) %&gt;% head(5)
```

```
## # A tibble: 5 × 4
##   species  height  mass birth_year
##   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
## 1 Aleena       79    15        NaN
## 2 Besalisk    198   102        NaN
## 3 Cerean      198    82         92
## 4 Chagrian    196   NaN        NaN
## 5 Clawdite    168    55        NaN
```

---
# Joining operations

Central feature of the `dplyr` package invovles merging data from multiple tables with [join operations](https://cran.r-project.org/web/packages/dplyr/vignettes/two-table.html).
- `inner_join(df1, df2)`
- `left_join(df1, df2)`
- `right_join(df1, df2)`
- `full_join(df1, df2)`
- `semi_join(df1, df2)`
- `anti_join(df1, df2)`

---
# Joining operations

- For some simple examples, we'll need some data sets that come bundled with the [nycflights13 package](http://github.com/hadley/nycflights13). 

- Load it now and then inspect these data frames in your own console.



```r
library(nycflights13)
flights 
planes
```

---
# Left join

Let's perform a [left join](https://stat545.com/bit001_dplyr-cheatsheet.html#left_joinsuperheroes-publishers) on the flights and planes datasets. 

--


```r
left_join(flights, planes) %&gt;%
  select(year, month, day, dep_time, arr_time, carrier, flight, tailnum, type, model)  %&gt;% head(5)
```

```
## # A tibble: 5 × 10
##    year month   day dep_time arr_time carrier flight tailnum type  model
##   &lt;int&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;
## 1  2013     1     1      517      830 UA        1545 N14228  &lt;NA&gt;  &lt;NA&gt; 
## 2  2013     1     1      533      850 UA        1714 N24211  &lt;NA&gt;  &lt;NA&gt; 
## 3  2013     1     1      542      923 AA        1141 N619AA  &lt;NA&gt;  &lt;NA&gt; 
## 4  2013     1     1      544     1004 B6         725 N804JB  &lt;NA&gt;  &lt;NA&gt; 
## 5  2013     1     1      554      812 DL         461 N668DN  &lt;NA&gt;  &lt;NA&gt;
```

---
# Left join

`dplyr` guessed about which columns to join on (i.e. columns that share the same name). It also told us its choices: 

&lt;br&gt;

```r
## Joining, by = c("year", "tailnum")
```

--
&lt;br&gt;
Problem: the variable "year" does not have a consistent meaning across our joining datasets!
- *year of flight* versus *year of construction*

---
# Left join

Luckily, there's an easy way to avoid this problem. 
- See if you can figure it out before turning to the next slide.
- Try `?dplyr::join`.

---
# Left join

Let's be more explicit with the `by = ` argument:

```r
left_join(
  flights,
  planes %&gt;% rename(year_built = year), ## Not necessary w/ below line, but helpful
  by = "tailnum" ## Be specific about the joining column
  ) %&gt;%
  select(year, month, day, dep_time, arr_time, carrier, flight, tailnum, year_built, type, model) %&gt;%
  head(5) ## Just to save vertical space on the slide
```

```
## # A tibble: 5 × 11
##    year month   day dep_time arr_time carrier flight tailnum year_built type    
##   &lt;int&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;   
## 1  2013     1     1      517      830 UA        1545 N14228        1999 Fixed w…
## 2  2013     1     1      533      850 UA        1714 N24211        1998 Fixed w…
## 3  2013     1     1      542      923 AA        1141 N619AA        1990 Fixed w…
## 4  2013     1     1      544     1004 B6         725 N804JB        2012 Fixed w…
## 5  2013     1     1      554      812 DL         461 N668DN        1991 Fixed w…
## # … with 1 more variable: model &lt;chr&gt;
```

---
# Be specific

What happens if we again specify the join column but don't rename the ambiguous "year"? 

```r
left_join(flights, planes, ## Not renaming "year" to "year_built" this time
  by = "tailnum") %&gt;%
  select(contains("year"), month, day, dep_time, arr_time, carrier, flight, tailnum, type, model) %&gt;%
  head(5)
```

```
## # A tibble: 5 × 11
##   year.x year.y month   day dep_time arr_time carrier flight tailnum type  model
##    &lt;int&gt;  &lt;int&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;
## 1   2013   1999     1     1      517      830 UA        1545 N14228  Fixe… 737-…
## 2   2013   1998     1     1      533      850 UA        1714 N24211  Fixe… 737-…
## 3   2013   1990     1     1      542      923 AA        1141 N619AA  Fixe… 757-…
## 4   2013   2012     1     1      544     1004 B6         725 N804JB  Fixe… A320…
## 5   2013   1991     1     1      554      812 DL         461 N668DN  Fixe… 757-…
```

--

What is "year.x" vs "year.y"?

---
# Other dplyr functions

`pull()`: Extract a column from a data frame as a vector or scalar.
- e.g. `starwars %&gt;% filter(gender=="female") %&gt;% pull(height)`

--

`count()` and `distinct()`: Number and isolate unique observations.
- e.g. `starwars %&gt;% count(species)`, or `starwars %&gt;% distinct(species)`
- You could also use a combination of `mutate()`, `group_by()`, and `n()`, e.g. `starwars %&gt;% group_by(species) %&gt;% mutate(num = n())`
- Built-in combination using `add_count()`.


---
# Other dplyr functions

There is also a whole class of [window functions](https://cran.r-project.org/web/packages/dplyr/vignettes/window-functions.html) for getting leads and lags, ranking, creating cumulative aggregates, etc.

See `vignette("window-functions")` for more.

---
# Other summaries

- Summary statistics can get cumbersome with `summarize()`
- Quick summary of selected variables: `sumtable()` from the `vtable` package

---
# Some dplyr tips

- Any `group_by()` statement stays until `ungroup()`

--

- Look out for `plyr` package. Do not use `plyr` and `dplyr` together. Just don't do it.

.center[
  ![:scale 800px](https://media.giphy.com/media/3owzVQ23NzdKvVALUA/giphy.gif)
]


---
class: inverse, center, middle
name: tidyr

# tidyr

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;

---
# Key tidyr verbs

1. `pivot_wider()` and `pivot_longer()` to reshape data between wide and long format

2. `separate()`: Split one column into multiple columns

3. `unite()`: Combine multiple columns into one

---
# 1) tidyr::pivot_longer()


```r
stocks &lt;- tibble(
  time = as.Date('2009-01-01') + 0:1,
  X = rnorm(2, 0, 1),
  Y = rnorm(2, 0, 2),
  Z = rnorm(2, 0, 4)
  )
stocks
```

```
## # A tibble: 2 × 4
##   time            X      Y      Z
##   &lt;date&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 2009-01-01  0.192 -1.62  -0.780
## 2 2009-01-02 -0.907 -0.332  6.00
```
---
# 1) tidyr::pivot_longer()

```r
tidy_stocks &lt;- stocks %&gt;% 
  pivot_longer(cols=c("X","Y","Z"), 
               names_to="stock", values_to="price")
tidy_stocks
```

```
## # A tibble: 6 × 3
##   time       stock  price
##   &lt;date&gt;     &lt;chr&gt;  &lt;dbl&gt;
## 1 2009-01-01 X      0.192
## 2 2009-01-01 Y     -1.62 
## 3 2009-01-01 Z     -0.780
## 4 2009-01-02 X     -0.907
## 5 2009-01-02 Y     -0.332
## 6 2009-01-02 Z      6.00
```

---
# 1) tidyr::pivot_longer()

### Aside: Remembering the syntax 

There's a long-running joke about no-one being able to remember Stata's "reshape" command. ([Exhibit A](https://twitter.com/scottimberman/status/1036801308785864704).)

It's easy to see this happening with `pivot_wider()` and `pivot_longer()` too. 

---
# 1) tidyr::pivot_wider()


```r
tidy_stocks %&gt;% 
  pivot_wider(values_from="price",names_from="stock")
```

```
## # A tibble: 2 × 4
##   time            X      Y      Z
##   &lt;date&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 2009-01-01  0.192 -1.62  -0.780
## 2 2009-01-02 -0.907 -0.332  6.00
```

---
# 2) tidyr::separate()


```r
economists &lt;- tibble(
  name = c("Abhijit Banerjee", "Esther Duflo", "Michael Kremer")
)
economists
```

```
## # A tibble: 3 × 1
##   name            
##   &lt;chr&gt;           
## 1 Abhijit Banerjee
## 2 Esther Duflo    
## 3 Michael Kremer
```

---
# 2) tidyr::separate()

```r
economists %&gt;% separate(name, c("first_name", "last_name")) 
```

```
## # A tibble: 3 × 2
##   first_name last_name
##   &lt;chr&gt;      &lt;chr&gt;    
## 1 Abhijit    Banerjee 
## 2 Esther     Duflo    
## 3 Michael    Kremer
```

--

Should also specify the separation character with `separate(..., sep=" ")`.

---
# 3) tidyr::separate_rows()

A related function is `separate_rows()` for splitting into new rows

```r
jobs &lt;- tibble(
  name = c("Jack", "Jill"),
  occupation = c("Homemaker", "Philosopher, Philanthropist, Troublemaker") 
  ) 
jobs
```

```
## # A tibble: 2 × 2
##   name  occupation                               
##   &lt;chr&gt; &lt;chr&gt;                                    
## 1 Jack  Homemaker                                
## 2 Jill  Philosopher, Philanthropist, Troublemaker
```
---
# 3) tidyr::separate_rows()


```r
## Now split out Jill's various occupations into different rows
jobs %&gt;% separate_rows(occupation)
```

```
## # A tibble: 4 × 2
##   name  occupation    
##   &lt;chr&gt; &lt;chr&gt;         
## 1 Jack  Homemaker     
## 2 Jill  Philosopher   
## 3 Jill  Philanthropist
## 4 Jill  Troublemaker
```
---
# 4) tidyr::unite()


```r
gdp &lt;- data.frame(
  yr = rep(2016, times = 4),
  mnth = rep(1, times = 4),
  dy = 1:4,
  gdp = rnorm(4, mean = 100, sd = 2)
  )
gdp 
```

```
##     yr mnth dy       gdp
## 1 2016    1  1  98.72684
## 2 2016    1  2  98.91063
## 3 2016    1  3  99.23657
## 4 2016    1  4 105.45198
```
---
# 4) tidyr::unite()

```r
## Combine "yr", "mnth", and "dy" into one "date" column
gdp %&gt;% unite(date, c("yr", "mnth", "dy"), sep = "-")
```

```
##       date       gdp
## 1 2016-1-1  98.72684
## 2 2016-1-2  98.91063
## 3 2016-1-3  99.23657
## 4 2016-1-4 105.45198
```

---
# 4) tidyr::unite()

`unite()` automatically creates character variable:

```r
gdp_u &lt;- gdp %&gt;% unite(date, c("yr", "mnth", "dy"), sep = "-") %&gt;% as_tibble()
gdp_u
```

```
## # A tibble: 4 × 2
##   date       gdp
##   &lt;chr&gt;    &lt;dbl&gt;
## 1 2016-1-1  98.7
## 2 2016-1-2  98.9
## 3 2016-1-3  99.2
## 4 2016-1-4 105.
```

--

Use `mutate()` with [lubridate](https://lubridate.tidyverse.org/) date functions to change the variable type.

---

# 4) tidyr::unite()


```r
library(lubridate)
gdp_u %&gt;% mutate(date = ymd(date))
```

```
## # A tibble: 4 × 2
##   date         gdp
##   &lt;date&gt;     &lt;dbl&gt;
## 1 2016-01-01  98.7
## 2 2016-01-02  98.9
## 3 2016-01-03  99.2
## 4 2016-01-04 105.
```

---
# Other tidyr functions

- `drop_na()` to drop missing values among specified columns

- `fill()` to impute missing values from past/future values

- `replace_na()` to replace missing values with known value

---
# Summary

.pull-left[
### dplyr
1. `filter()`
2. `arrange()`
3. `select()`
4. `mutate()`
5. `summarise()`
]
.pull-right[
### tidyr
1. `pivot_longer()`
2. `pivot_wider()`
3. `separate()`
4. `unite()`
]

--
&lt;/br&gt;
Other useful items include: pipes (`%&gt;%`), grouping (`group_by()`), joining functions (`left_join()`, `inner_join`, etc.).



&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: real_data

# Real World

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;


---
# Practice data versus the real world

&lt;br&gt;
.center[
  ![:scale 800px](https://media.giphy.com/media/3oz8xGme7vEndhrsly/giphy.gif)
]

---
# Advice 1: Be patient and careful in your coding

.center[
  ![](https://media.giphy.com/media/fsQJBs0rtxpYoN4m0J/giphy.gif)
]

---
# Advice 2: Comment, comment, comment

You don't want to end up like this guy...

.center[
  ![:scale 600px](https://media.giphy.com/media/S7u66urzxc2J2/giphy.gif)
]

&lt;!-- New Section --&gt;
---
class: inverse, center, middle
name: ma_data

# Medicare Advantage

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=1055px&gt;&lt;/html&gt;


---
# Medicare Advantage 

Let's work with the [Medicare Advantage GitHub repository](https://github.com/imccart/Medicare-Advantage)


---
# Access the data

First step is to download the raw data that we'll be using, or work in AWS:
 - [Monthly Enrollment](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/MCRAdvPartDEnrolData/Monthly-Enrollment-by-Contract-Plan-State-County.html)
 - [Plan Characteristics](https://www.cms.gov/Medicare/Prescription-Drug-Coverage/PrescriptionDrugCovGenIn/index)
 - [Service Areas](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/MCRAdvPartDEnrolData/MA-Contract-Service-Area-by-State-County.html)

--

Lots more out there, but this is enough for now.

---
# 1) Contract/enrollment info

```r
for (y in 2006:2015) {
  monthlist=get(paste0("monthlist_",y))
  for (m in monthlist) {
    ## Basic contract/plan information
*   ma.path=paste0("...CPSC_Contract_Info_",y,"_",m,".csv")
    contract.info=read_csv(ma.path,
                           skip=1,
                           col_names = c("contractid","planid","org_type","plan_type",
                                         "partd","snp","eghp","org_name","org_marketing_name",
                                         "plan_name","parent_org","contract_date"),
                           col_types = cols(
                             contractid = col_character(),
                             planid = col_double(),
                             ...
                           ))
```

???
This imports the contract information (name, insurer, plan identifiers) into an object `contract.info` from the filepath in `ma.path`.

---
# 1) Contract/enrollment info

```r
    ## Clean the contract level data
    contract.info = contract.info %&gt;%
      group_by(contractid, planid) %&gt;%
      mutate(id_count=row_number())
    
    contract.info = contract.info %&gt;%
      filter(id_count==1) %&gt;%
      select(-id_count)
```

???
This reduces the data to unique contract and plan IDs. Note the `row_number()` to assign a running count for each pair.

---
# 1) Contract/enrollment info

```r
    ## Enrollments per plan
*   ma.path=paste0("...CPSC_Enrollment_Info_",y,"_",m,".csv")
    enroll.info=read_csv(ma.path,
                         skip=1,
                         col_names = c("contractid","planid","ssa","fips","state","county","enrollment"),
                         col_types = cols(
                           contractid = col_character(),
                           planid = col_double(),
                           ssa = col_double(),
                           fips = col_double(),
                           state = col_character(),
                           county = col_character(),
                           enrollment = col_double()
                         ),na="*")
```

???
This imports the enrollment data (count of enrollees) into an object `enroll.info` from the filepath `ma.path`

---
# 1) Contract/enrollment info

```r
    ## Merge contract info with enrollment info
    plan.data = contract.info %&gt;%
      left_join(enroll.info, by=c("contractid", "planid")) %&gt;%
      mutate(month=as.numeric(m),year=y)
    
    assign(paste0("plan.data.",m),plan.data)
  }
```

???
This merges the contract details with enrollments. Note that this is for each year and month. `assign()` is not good coding practice but I'm not sure of a clearly better way in this case.

---
# 1) Contract/enrollment info

```r
  ## Append monthly enrollment info for each year
  if (y==2006) {
    plan.month=rbind(plan.data.07, plan.data.08, plan.data.09, plan.data.10,
                     plan.data.11, plan.data.12)
  } else {
    plan.month=rbind(plan.data.01, plan.data.02, plan.data.03, plan.data.04,
                     plan.data.05, plan.data.06, plan.data.07, plan.data.08, 
                     plan.data.09, plan.data.10, plan.data.11, plan.data.12)
  }
```

???
This creates a final `plan.month` object consisting of the appended (or stacked) monthly `plan.data` objects.

---
# 1) Contract/enrollment info

```r
  ## Fill in missing fips codes (by state and county)
  plan.month = plan.month %&gt;%
    group_by(state, county) %&gt;%
    fill(fips)

  ## Fill in missing plan characteristics by contract and plan id
  plan.month = plan.month %&gt;%
    group_by(contractid, planid) %&gt;%
    fill(plan_type, partd, snp, eghp, plan_name)
  
  ## Fill in missing contract characteristics by contractid
  plan.month = plan.month %&gt;%
    group_by(contractid) %&gt;%
    fill(org_type,org_name,org_marketing_name,parent_org)
```

???
Some cleanup by filling in missing values using "nearby" values

---
# 1) Contract/enrollment info

```r
  ## Collapse from monthly data to yearly
  plan.year = plan.month %&gt;%
    group_by(contractid, planid, fips) %&gt;%
    arrange(contractid,planid,fips,month) %&gt;%
    summarize(avg_enrollment=mean(enrollment),sd_enrollment=sd(enrollment),
              min_enrollment=min(enrollment),max_enrollment=max(enrollment),
              first_enrollment=first(enrollment),last_enrollment=last(enrollment),
              state=last(state),county=last(county),org_type=last(org_type),
              plan_type=last(plan_type),partd=last(partd),snp=last(snp),
              eghp=last(eghp),org_name=last(org_name),org_marketing_name=last(org_marketing_name),
              plan_name=last(plan_name),parent_org=last(parent_org),contract_date=last(contract_date),
              year=last(year))
    
  write_rds(plan.year,paste0(path.data.final,"/ma_data_",y,".rds"))
}
```

???
Summarize monthly data to the yearly level, and save data object as `ma_data_` (year) at filepath `path.data.final`

---
# 1) Contract/enrollment info

```r
full.ma.data &lt;- readRDS(paste0(path.data.final,"/ma_data_2006.rds"))
for (y in 2007:2015) {
  full.ma.data &lt;- rbind(full.ma.data, paste0(path.data.final,"/ma_data_",y,".rds"))
}
```

???

Appends yearly data into one large file, `full.ma.data`

---
# All together now

Now let's move to AWS and try some of this together.
1. Clone repository
2. Create your own repository
3. Remove remote, `git remote rm origin`
4. Set new remote `git remote add origin NEW_URL` (where NEW_URL is your own repository)
5. git push origin HEAD:master
6. git push --set-upstream origin master

---
# All together now

Next, you'll need to make some changes to get things to work:

1. Create new `paths.R` script with:

```r
path.code="R_code"
path.data.ma="/data/medicare-advantage"
path.data.ffs="/data/medicare-ffs"
```

2. Update file paths

3. Remove "path.data.final" and rely on `here()` package

---
# Number of plans available



```r
full.ma.data %&gt;% group_by(fips, year) %&gt;% select(fips, year) %&gt;% summarize(plan_count=n()) %&gt;%
  ggplot(aes(x=as.factor(year),y=plan_count)) + 
  stat_summary(fun.y="mean", geom="bar") +
  labs(
    x="Year",
    y="Number of Plans",
    title="Average Number of Plans per County"
  ) + scale_y_continuous(labels=comma) +
  theme_bw()
```

.plot-callout[
&lt;img src="00-introduction_files/figure-html/all-plans-callout-1.png" style="display: block; margin: auto;" /&gt;
]


---
# Number of plans available

&lt;img src="00-introduction_files/figure-html/all-plans-output-1.png" style="display: block; margin: auto;" /&gt;


---
# Number of plans available

.left-code[

```r
full.ma.data %&gt;% 
  group_by(fips, year) %&gt;% 
  select(fips, year) %&gt;% 
  summarize(plan_count=n()) %&gt;%
  ggplot(aes(x=as.factor(year),y=plan_count)) + 
  stat_summary(fun.y="mean", geom="bar") +
  labs(x="Year", y="Number of Plans", title="Average Number of Plans per County") +
  scale_y_continuous(labels=comma) +
  theme_bw()
```
]

.right-plot[
![](00-introduction_files/figure-html/plot1-1.png)
]

---
# Number of plans available

.left-code[

```r
full.ma.data %&gt;% 
* filter(snp=="No" &amp; eghp=="No") %&gt;%
  group_by(fips, year) %&gt;% 
  select(fips, year) %&gt;% 
  summarize(plan_count=n()) %&gt;%
  ggplot(aes(x=as.factor(year),y=plan_count)) + 
  stat_summary(fun.y="mean", geom="bar") +
  labs(x="Year", y="Number of Plans", title="Average Number of Plans per County") +
  scale_y_continuous(labels=comma) +
  theme_bw()
```
]

.right-plot[
![](00-introduction_files/figure-html/plot2-1.png)
]



---
# Number of plans available

.left-code[

```r
full.ma.data %&gt;% 
  filter(snp=="No" &amp; eghp=="No") %&gt;%
* filter(planid &lt; 800 | planid &gt;= 900) %&gt;%
* filter(!is.na(planid)) %&gt;%
  group_by(fips, year) %&gt;% 
  select(fips, year) %&gt;% 
  summarize(plan_count=n()) %&gt;%
  ggplot(aes(x=as.factor(year),y=plan_count)) + 
  stat_summary(fun.y="mean", geom="bar") +
  labs(x="Year", y="Number of Plans", title="Average Number of Plans per County") +
  scale_y_continuous(labels=comma) +
  theme_bw()
```
]

.right-plot[
![](00-introduction_files/figure-html/plot3-1.png)
]


---
# Number of plans available

.left-code[

```r
full.ma.data %&gt;% 
  filter(snp=="No" &amp; eghp=="No") %&gt;%
  filter(planid &lt; 800 | planid &gt;= 900) %&gt;%
  filter(!is.na(planid)) %&gt;%
* inner_join(service.area %&gt;%
*              select(contractid, fips, year),
*            by=c("contractid", "fips", "year")) %&gt;%
  group_by(fips, year) %&gt;% 
  select(fips, year) %&gt;% 
  summarize(plan_count=n()) %&gt;%
  ggplot(aes(x=as.factor(year),y=plan_count)) + 
  stat_summary(fun.y="mean", geom="bar") +
  labs(x="Year", y="Number of Plans", title="Average Number of Plans per County") +
  scale_y_continuous(labels=comma) +
  theme_bw()
```
]

.right-plot[
![](00-introduction_files/figure-html/plot4-1.png)
]



---
# Interactive plot


&lt;iframe src="plan-plot.html" width="800" height="500" scrolling="yes" seamless="seamless" frameBorder="0"&gt; &lt;/iframe&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
